crates/sui/src/upgrade_compatibility.rs

```diff
@@ -513,85 +572,722 @@ fn compare_packages(
-            "{declaration_kind} '{identifier_name}' is missing",
-            declaration_kind = declaration_kind,
-            identifier_name = identifier_name,
-        )),
-        std::iter::empty::<(Loc, String)>(),
-        vec![format!(
-            "{declaration_kind} is missing expected {declaration_kind} '{identifier_name}', but found none",
+    diags.add(
+        Diagnostic::new(
+            Declarations::PublicMissing,
+            (loc, format!(
+                "{declaration_kind} '{identifier_name}' is missing",
+                declaration_kind = declaration_kind,
+                identifier_name = identifier_name,
+            )),
+            std::iter::empty::<(Loc, String)>(),
+            vec![format!(
+                "{declaration_kind} is missing expected {declaration_kind} '{identifier_name}', but found none",
+            ),
+                 format!(
+                     "{declaration_kind}s are part of a module's public interface and cannot be removed or changed during an upgrade",
+                 ),
+                 format!(
+                     "add missing {declaration_kind} '{identifier_name}' back to the module '{module_name}'.",
+                 )],
+        )
+    );
+
+    Ok(diags)
+}
+
+/// Return a diagnostic for a function signature mismatch.
+/// start by checking the lengths of the parameters and returns and return a diagnostic if they are different
+/// if the lengths are the same check each parameter piece wise and return a diagnostic for each mismatch
+fn function_signature_mismatch_diag(
+    function_name: &Identifier,
+    old_function: &Function,
+    new_function: &Function,
+    compiled_unit_with_source: &CompiledUnitWithSource,
+    lookup: &IdentifierTableLookup,
+) -> Result<Diagnostics, Error> {
+    let mut diags = Diagnostics::new();
+
+    let old_func_index = lookup
+        .function_identifier_to_index
+        .get(function_name)
+        .context("Unable to get function index")?;
+
+    let new_func_sourcemap = compiled_unit_with_source
+        .unit
+        .source_map
+        .get_function_source_map(FunctionDefinitionIndex::new(*old_func_index))
+        .context("Unable to get function source map")?;
+
+    let def_loc = new_func_sourcemap.definition_location;
+
+    // handle function arguments
+    if old_function.parameters.len() != new_function.parameters.len() {
+        diags.add(
+            Diagnostic::new(
+                Function_::SignatureMismatch,
+                (def_loc, format!("Expected {} parameters, have {}", old_function.parameters.len(), new_function.parameters.len())),
+                Vec::<(Loc, String)>::new(),
+                vec![
+                    "Functions are part of a module's public interface and cannot be changed during an upgrade.".to_string(),
+                    format!("Restore the original function's parameters for function '{function_name}', expected {} parameters.", old_function.parameters.len())
```

Comment:
You might want to do some manual line breaks in the string here for some manual formatting, e.g.
```suggestion
                    format!("Restore the original function's parameters for \
                        function '{function_name}', expected {} parameters.", old_function.parameters.len())
```